subprojects/core/src/integTest/groovy/org/gradle/api/tasks/NestedInputIntegrationTest.groovy

```diff
@@ -0,0 +1,227 @@
+                bean = new NestedBeanWithInput(input: newInput${fileType}())
+                bean.input.set(generator.outputFile)
+            }
+        """
+
+        when:
+        run 'consumer'
+
+        then:
+        executedAndNotSkipped(':generator', ':consumer')
+
+        where:
+        fileType    | type
+        'File'      | RegularFileProperty
+        'Directory' | DirectoryProperty
+    }
+
+    def "Nested FileCollection input adds a task dependency"() {
+        buildFile << """
+            class TaskWithNestedProperty extends DefaultTask  {
+                @Nested
+                Object bean
+            }
+            
+            class NestedBeanWithInput {
+                @InputFiles
+                FileCollection input
+            }
+            
+            class TaskGeneratingFile extends DefaultTask {
+                @OutputFile
+                RegularFileProperty outputFile = newOutputFile()
+                
+                @TaskAction
+                void doStuff() {
+                    outputFile.getAsFile().get().text = "Hello"
+                }
+            }
+            
+            task generator(type: TaskGeneratingFile) {
+                outputFile.set(project.layout.buildDirectory.file('output'))
+            }
+            
+            task consumer(type: TaskWithNestedProperty) {
+                bean = new NestedBeanWithInput(input: files(generator.outputFile))
+            }
+        """
+
+        when:
+        run 'consumer'
+
+        then:
+        executedAndNotSkipped(':generator', ':consumer')
+    }
+
+    @ToBeImplemented
+    def "Nested input using output file property of different task adds a task dependency"() {
+        buildFile << """
+            class TaskWithNestedProperty extends DefaultTask  {
+                @Nested
+                Object bean
+            }
+            
+            class NestedBeanWithInput {
+                @InputFile
+                RegularFileProperty file
+            }
+            
+            class TaskGeneratingFile extends DefaultTask {
+                @OutputFile
+                RegularFileProperty outputFile = newOutputFile()
+                
+                @TaskAction
+                void doStuff() {
+                    outputFile.getAsFile().get().text = "Hello"
+                }
+            }
+            
+            task generator(type: TaskGeneratingFile) {
+                outputFile.set(project.layout.buildDirectory.file('output'))
+            }
+            
+            task consumer(type: TaskWithNestedProperty) {
+                bean = new NestedBeanWithInput(file: generator.outputFile)
+            }
+        """
+
+        when:
+        run 'consumer'
+
+        then:
+        // Should have been executed
+        notExecuted(':generator')
+        // Should have been executed
+        skipped(':consumer')
+    }
+
+    def "Changing nested inputs during execution time is detected"() {
```

Comment:
I think we'd need a test like this for changing input files and output files, too.

Reply:
I changed the test, so it also tests for changes to input files and output files which are part of a nested bean.